command/certificate/p12.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/smallstep/cli/ui"
+	"github.com/smallstep/cli/utils"
+	"github.com/urfave/cli"
+
+	"software.sslmate.com/src/go-pkcs12"
+)
+
+func p12Command() cli.Command {
+	return cli.Command{
+		Name:      "p12",
+		Action:    command.ActionFunc(p12Action),
+		Usage:     `package a certificate and keys into a .p12 file`,
+		UsageText: `step certificate p12 <p12_path> [<crt_path>] [<key_path>] [**--ca**=<ca_crt_path>] [**--password**=<password>]`,
+		Description: `**step certificate p12** creates a .p12 (PFX / PKCS12)
+		file containing certificates and keys. This can then be used to import
+		into Windows / Firefox / Java applications.
+
+## EXIT CODES
+
+This command returns 0 on success and \>0 if any error occurs.
+
+## EXAMPLES
+
+Package a certificate and private key together:
+
+'''
+$ step certificate p12 foo.p12 foo.crt foo.key
+'''
+
+Package a certificate and private key together, and include an intermediate certificate:
+
+'''
+$ step certificate p12 foo.p12 foo.crt foo.key --ca intermediate.crt
+'''
+
+Package a CA certificate into a "trust store" for Java applications
+
+'''
+$ step certificate p12 trust.p12 --ca ca.crt
+'''
+`,
+		Flags: []cli.Flag{
+			cli.StringSliceFlag{
+				Name: "ca",
+				Usage: `Add a CA or intermediate certificate to the .p12 file. Use the '--ca'
+flag multiple times to add multiple CAs or intermediates.`,
+			},
+			cli.StringFlag{
+				Name:  "password",
+				Usage: `Set the password to encrypt the .p12 file.`,
+			},
+			flags.Force,
+		},
+	}
+}
+
+func p12Action(ctx *cli.Context) error {
+	if err := errs.MinMaxNumberOfArguments(ctx, 1, 3); err != nil {
+		return err
+	}
+
+	p12File := ctx.Args().Get(0)
+	crtFile := ctx.Args().Get(1)
+	keyFile := ctx.Args().Get(2)
+	caFiles := ctx.StringSlice("ca")
+	password := ctx.String("password")
+	hasKeyAndCert := crtFile != "" && keyFile != ""
+
+	//If either key or cert are provided, both must be provided
+	if !hasKeyAndCert && (crtFile != "" || keyFile != "") {
+		return errs.MissingArguments(ctx, "key_file")
+	}
+
+	//If no key and cert are provided, ca files must be provided
+	if !hasKeyAndCert && len(caFiles) == 0 {
+		return errors.Errorf("flag '--%s' must be provided when no <crt_path> and <key_path> are present", "ca")
+	}
+
+	x509CAs := []*x509.Certificate{}
+	for _, caFile := range caFiles {
+		x509CA, err := pemutil.ReadCertificate(caFile)
+		if err != nil {
+			return errors.Wrap(err, "error reading CA certificate")
+		}
+		x509CAs = append(x509CAs, x509CA)
+	}
+
+	if password == "" {
+		pass, err := ui.PromptPassword("Please enter a password to encrypt the .p12 file")
+		if err != nil {
+			return errors.Wrap(err, "error reading password")
+		}
+		password = string(pass)
+	}
+
+	var pkcs12Data []byte
+	var err error
+
+	if hasKeyAndCert {
+		//If we have a key and certificate, we're making an identity store
+		x509Cert, err := pemutil.ReadCertificate(crtFile)
+		if err != nil {
+			return errors.Wrap(err, "error reading certificate")
+		}
```

Comment:
An X.509 certificate is usually bundled with an intermediary, we do that by default if you are using `step-ca`. I think it makes sense that if we pass a bundled certificate, we should include the intermediary automatically in the list of CAs. Don't you think?
You can use `pemutil.ReadCertificateBundle(filename string) ([]*x509.Certificate, error)` to get all certificates.

Reply:
What would happen if a CA is added multiple times? For example one from the cert and one from the `--ca` flag.

Reply:
I'm not sure - I suspect it will be fine as PKCS12 is a dumb collection of certificates. I'll build a few different types and do some tests.

Reply:
Regarding bundles - if a bundle is passed to the `--ca` option it's easy - append all the certs to the CA list.
If one is passed to `<crt_file>`, we need to work out which one the key applies to, and which is the intermediate as they get stored in different places in the result.

Is the best way to do this extract the public key from the certs and the private key, and compare to find a match? I had a look around the codebase but I couldnt see any existing code that does such a check.

Reply:
A X.509 certificate is usually a bundle like:
```
-----BEGIN CERTIFICATE-----
MII.... the server certificate
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MII.... the intermediate certificate
-----END CERTIFICATE-----
```

It can also include multiple intermediates at the end. For example, in TLS this is required because your browser only knows the root certificates and he needs the intermediate to build the validation path.

Check for example:
```sh
step certificate inspect --format pem --bundle https://www.google.com/
```

So using a PKI with the structure `root`->`intermediate`->`leaf` and using a CA like `step-ca` it will make sense to create pkcs#12 certificate like:

```sh
step ca certificate mariano@smallstep.com mariano.crt mariano.key
step certificate p12 mariano.crt mariano.key --ca $(step path)/certs/root_ca.crt
```

Reply:
The server cert is always the first. The list of intermediates are in order at the end.